src/dbnode/storage/database.go

```diff
@@ -317,11 +309,54 @@ func (d *db) UpdateOwnedNamespaces(newNamespaces namespace.Map) error {
 			"restart the process if you want changes to take effect")
 	}
 
-	// enqueue bootstraps if new namespaces
 	if len(adds) > 0 {
-		d.queueBootstrapWithLock()
+		d.Lock()
+		defer d.Unlock()
+		// if db was not yet bootstrapped, it is safe to add namespaces immediately
+		if d.bootstraps == 0 {
+			return d.addNamespacesWithLock(adds)
+		}
+
+		// if mediator is not opened, it is safe to add namespaces immediately
+		if mediator := d.mediator; mediator == nil || !mediator.IsOpen() {
+			if err := d.addNamespacesWithLock(adds); err != nil {
+				enrichedErr := fmt.Errorf("unable to add namespaces: %w", err)
+				d.log.Error(enrichedErr.Error())
+				return err
+			}
+			d.enqueueBootstrapWithLock()
+			return nil
+		}
+
+		if err := d.mediator.EnqueueMutuallyExclusiveFn(func() {
+			if err := d.addNamespaces(adds); err != nil {
+				instrument.EmitAndLogInvariantViolation(d.opts.InstrumentOptions(),
+					func(l *zap.Logger) {
+						l.Error("failed to add namespaces", zap.Error(err))
+					})
+			}
+		}); err != nil {
+			// should not happen.
+			instrument.EmitAndLogInvariantViolation(d.opts.InstrumentOptions(),
+				func(l *zap.Logger) {
+					l.Error("failed to enqueue addNamespaces fn", zap.Error(err))
+				})
+		}
 	}
+	return nil
+}
 
+func (d *db) addNamespaces(namespaces []namespace.Metadata) error {
+	d.Lock()
+	defer d.Unlock()
+	// add any namespaces marked for addition
+	if err := d.addNamespacesWithLock(namespaces); err != nil {
+		enrichedErr := fmt.Errorf("unable to add namespaces: %w", err)
+		d.log.Error(enrichedErr.Error())
+		return err
```

Comment:
To match zap logging style, this should just be:
```suggestion
		d.log.Error("unable to add namespaces", zap.Error(err))
		return err
```